lib/Hal.js

```diff
@@ -0,0 +1,67 @@
+
+class Hal {
+  constructor(phrase) {
+    if (!phrase) throw 'A phrase is required';
+    this.phrase = phrase
+  }
+
+  listen(phrase) {
+    this.phrase = phrase;
+    switch (true) {
+      /* checks if it starts with 'Do' */
+      case /^Do.*/i.test(this.phrase):
+        if (/you/i.test(this.phrase)) {
+          if (/want.*this.*/i.test(this.phrase)) {
+            return 'Is it free?'
+          }
+        }
+      case /.*open.+door.*/i.test(this.phrase):
+        return `I'm sorry, Dave. I'm afraid I can't do that.`
+
+      case /Good.*/i.test(this.phrase):
+        if (/.*morning.*/) {
+          return 'Good Morning Dave'
+        }
+
+      /* checks if starts with 'what'*/
+      case /^what.*/i.test(this.phrase):
+        if (/.*do.*/i.test(this.phrase)) {
+          if (/.*want.*/i.test(this.phrase)) {
+            if (/.*dinner\W/i.test(this.phrase)) {
```

Comment:
This is actually more specific, as it forces those words to be in the correct order, where yours disregards order.